tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs

```diff
@@ -0,0 +1,313 @@
+            It.IsAny<int>(),
+            It.IsAny<CancellationToken>());
+
+    [Fact]
+    public async Task ChatsAreBeingLoadedInBatches()
+    {
+        var chats = new Mock<IChatsBackend>();
+        chats
+            .Setup(ListChangedCall)
+            .Returns<long, long, ChatId, int, CancellationToken>(
+                (minVersion, _, _, size, _) => GetNextBatch(minVersion, size)
+            );
+
+        const int batchSize = 5;
+
+        var sequence = new InfiniteChatSequence(Clock, chats.Object, Log) {
+            BatchSize = batchSize,
+        };
+
+        const int maxChatCount = 23;
+        var prevVersion = 0L;
+        await foreach (var (_, version) in sequence.LoadAsync(1, CancellationToken.None).Take(maxChatCount)) {
+            Assert.True(version > prevVersion);
+            prevVersion = version;
+        }
+
+        chats.Verify(
+            ListChangedCall,
+            Times.Exactly((maxChatCount + batchSize - 1) / batchSize)
+        );
+    }
+
+    [Fact]
+    public async Task ThereIsADelayBeforeRetryingInCaseOfAnEmptyBatch()
+    {
+        var emptyBatchDelay = TimeSpan.FromSeconds(11111);
+
+        var clock = new Mock<IMomentClock>();
+        clock
+            .Setup(x => x.Delay(It.IsAny<TimeSpan>(), It.IsAny<CancellationToken>()))
+            .Returns(Task.CompletedTask);
+
+        var batchCount = 0;
+        var chats = new Mock<IChatsBackend>();
+        chats
+            .Setup(ListChangedCall)
+            .Returns<long, long, ChatId, int, CancellationToken>(
+                (minVersion, _, _, size, _) => batchCount++ == 0
+                    ? Task.FromResult(new ApiArray<Chat.Chat>([]))
+                    : GetNextBatch(minVersion, size)
+            );
+
+        var sequence = new InfiniteChatSequence(clock.Object, chats.Object, Log) {
+            NoChatsIdleInterval = emptyBatchDelay
+        };
+
+        await sequence.LoadAsync(1, CancellationToken.None).Take(1).EnumerateAll();
+
+        chats.Verify(ListChangedCall, Times.Exactly(2));
+        clock.Verify(
+            x => x.Delay(It.Is<TimeSpan>(ts => ts==emptyBatchDelay), It.IsAny<CancellationToken>()), Times.Once);
+        clock.VerifyNoOtherCalls();
+    }
+
+    private static Task<ApiArray<Chat.Chat>> GetNextBatch(long lastVersion, int batchSize)
+    {
+        var batch = new Chat.Chat[batchSize];
+        for (var i = 0; i < batchSize; i++) {
+            var chatId = new ChatId(Generate.Option);
+            batch[i] = new Chat.Chat(chatId, lastVersion + i + 1);
+        }
+        return Task.FromResult(new ApiArray<Chat.Chat>(batch));
+    }
+
+    [Fact]
+    public async Task LoadMethodThrowsIfCancellationRequestedBeforeCall()
+    {
+        var sequence = new InfiniteChatSequence(Clock, Chats, Log);
+
+        var cancellationSource = new CancellationTokenSource();
+        await cancellationSource.CancelAsync();
+        var e = await Assert.ThrowsAsync<OperationCanceledException>(async () => {
```

Comment:
btw, there is an extension which convers Task/ValueTask to Func<Task> and allows to use fluent assertions in a fluent manner:
```
await sequence.LoadAsync(1, cancellationSource.Token).Take(1).EnumerateAll()
  .AsAsyncFunc().Should().ThrowAsync<OperationCancelledException>
```

Reply:
I would use this one, but I need the exception instance to run one more assertion in the next statement.